src/Storages/System/StorageSystemZooKeeper.cpp

```diff
@@ -443,8 +446,17 @@ static Paths extractPath(const ActionsDAG::NodeRawConstPtrs & filter_nodes, Cont
     for (const auto * node : filter_nodes)
         extractPathImpl(*node, res, context, allow_unrestricted);
 
+    auto node1 = res.find("/");
+    auto node2 = res.find("");
+    if ((node1 != res.end() && node1->second != ZkPathType::Exact) || (node2 != res.end() && node2->second != ZkPathType::Exact))
```

Comment:
With the old interpreter we have this:

```
EXPLAIN SYNTAX
SELECT
    name,
    path
FROM
(
    SELECT
        path,
        name
    FROM system.zookeeper
    WHERE path LIKE '/clickhouse/%'
    ORDER BY name ASC
)
WHERE path LIKE '%02221_system_zookeeper_unrestricted_like_test_00z68zd5/02221_system_zookeeper_unrestricted_like%'

Query id: 019e4581-a1fc-4428-8548-8749daa63dea

┌─explain───────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┐
│ SELECT                                                                                                                                                    │
│     name,                                                                                                                                                 │
│     path                                                                                                                                                  │
│ FROM                                                                                                                                                      │
│ (                                                                                                                                                         │
│     SELECT                                                                                                                                                │
│         path,                                                                                                                                             │
│         name                                                                                                                                              │
│     FROM system.zookeeper                                                                                                                                 │
│     WHERE (path LIKE '/clickhouse/%') AND (path LIKE '%02221_system_zookeeper_unrestricted_like_test_00z68zd5/02221_system_zookeeper_unrestricted_like%') │
│     ORDER BY name ASC                                                                                                                                     │
│ )                                                                                                                                                         │
│ WHERE path LIKE '%02221_system_zookeeper_unrestricted_like_test_00z68zd5/02221_system_zookeeper_unrestricted_like%'                                       │
└───────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┘

13 rows in set. Elapsed: 0.015 sec.
```

As the condition has been pushed inside the inner query, the filter '%02221_system_zookeeper_unrestricted_like_test_00z68zd5/02221_system_zookeeper_unrestricted_like%' is already allowing everything. In fact it happens twice. So there is 3 filters, 2 visiting everything and one visiting `/clickhouse/%`. If we know we are going to visit everything we can forget about everything else.